Various improvements and enhancements

NicoK opened the pull request:
This set of commits basically extends the usability of the current fraud detection demo to be used in benchmarks. It also fixes a few performance hogs such as logging all input elements and removes some dead code.
I hope the interaction with the web component still works - I always tried to adapt both projects but actually never ran the webapp.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/ververica/lab-fraud-detection/1)
<!-- Reviewable:end -->

NicoK committed:
add options for using discarding rules/alerts sinks

NicoK committed:
add and fix sink names

NicoK committed:
add a Rule generator with a static ruleset

This is mainly for testing only.

NicoK committed:
[hotfix] use diamond operator

NicoK committed:
[hotfix] no error logging if a rule is not yet updated in DynamicAlertFunction

NicoK committed:
[hotfix] modify a few log statements and don't use them for the normal workflow

- there is no need to print all elements that are being processed (and if there
  is, people can add that)
- log statements showing what rules have been processed/removed are demoted to
  TRACE now
- added some more details on (de)serialization failures

NicoK committed:
[hotfix] fix warnings reported by IntelliJ IDEA

NicoK committed:
tune log4j2 for slightly more condensed logs

+ remove the "rootlooger" entry

NicoK committed:
remove unused CLEAR_STATE_ALL_STOP control type

NicoK committed:
introduce sink parallelism to allow different patterns

Using "--sink-parallelism -1", for example, would chain all sinks into the main
processing operator and avoid unnecessary serialization.

NicoK committed:
[hotfix] transactionsPerSecond is only needed by TransactionsGenerator

NicoK committed:
tune gradle build file

NicoK committed:
upgrade Flink to 1.10

NicoK committed:
[cleanup] remove unused Rule#unique field

NicoK committed:
only change the restart strategy for local runs via "--local true"

NicoK committed:
[hotfix] add a proper dependency to jackson

(do not rely on Flink-shaded's version)

NicoK committed:
update gradlew and build.gradle to gradle 6.2

afedulov commented on flink-job/src/main/java/com/ververica/field/dynamicrules/RulesEvaluator.java:
```diff
@@ -158,22 +153,26 @@ public void run() throws Exception {
-            ? StreamExecutionEnvironment.createLocalEnvironmentWithWebUI(flinkConfig)
-            : StreamExecutionEnvironment.getExecutionEnvironment();
+    StreamExecutionEnvironment env;
+    if (isLocal) {
```
Probably `isLocal` was not the best name for this option. With this change there is a certain mix of two concerns together - the desire to run with a restart strategy that does not flood the logs if socket is not running and the desire not to start a WebUI in case you have something else running on the Flink UI port. If Flink will start the UI in a random, non occupied UI port and communicate this in the logs, I am for this change. If it causes failures, it makes sense to keep those parameters separate, but maybe better depict the intentions in their names.

NicoK replied:
I read this `local` parameter as a shortcut for local execution without worrying too much about the details (and I liked it because of that). I did not have the use case of something blocking your web UI port in my mind though.
In general, I would agree in separating these concerns, but I'm not sure we should follow that for this shortcut-parameter.
What do you think about this idea? Kust have a port(-range) as a parameter to `local` in the following way:

- ~~`local true`: defaults, with web-UI on default port, fixedDelayRestart (I would have liked to have this without parameters, but this is not possible without extending your wrappers)~~ (too hacky for now)
- `local -1`: defaults, no web-UI, fixedDelayRestart
- `local <portOrPortRange>`: defaults, web-UI on some port in the given range, fixedDelayRestart - the user has to find the port in the logs from this line:
`12:11:04.062 [main] INFO  o.a.f.r.d.DispatcherRestEndpoint - Web frontend listening at http://localhost:8081.`

afedulov replied:
What would be the option to get non-fixedDelayRestart and no local WebUI? Omitting this parameter, like in the current version?

NicoK replied:
FYI: finding this log line is currently not possible since too much is printed so that the output buffer is flushed too quickly (I addresses this in my performance improvements, so it will eventually be solved)

afedulov replied:
Yes, I know, that is a problem. I have this setup right now:https://github.com/afedulov/state-migration/blob/master/src/main/resources/log4j2.yaml

BTW, I have started using yaml-based log4j configs and can recommend them - they are much easier to work with.

NicoK replied:
yes, I would ignore non-fixedDelayRestart and no local WebUI - or actually, you can get this without setting `--local` in the first place (because then, nothing from that parameter is left)

Not sure setting Flink log level to error helps too much though as it also has some downsides during debugging / understanding Flink.

afedulov replied:
> I would ignore non-fixedDelayRestart and no local WebUI  

Not sure I get what you mean. I would like to keep this option for being able to run it on the platform, potentially for long time.

afedulov replied:
If just not setting the --local would get to no local UI and more resilient restarting strategy - this is perfectly fine.

NicoK replied:
ok, let me summarise the options in the current state:
- no `local`: just plain `StreamExecutionEnvironment.getExecutionEnvironment();` with no other changes (taking server defaults)
- `local -1`: `StreamExecutionEnvironment.getExecutionEnvironment();` (with no web-UI) but also setting fixedDelayRestart
- `local <portOrPortRange>`: `StreamExecutionEnvironment.createLocalEnvironmentWithWebUI()` with web-UI on some port in the given range, also setting fixedDelayRestart

I think, this covers all scenarios, most importantly not overriding the cluster defaults with no parameter (my original intention)

afedulov commented on flink-job/build.gradle:
```diff
@@ -103,29 +102,29 @@ dependencies {
-    compile "org.apache.logging.log4j:log4j-api:$log4jVersion"
-    compile "org.apache.logging.log4j:log4j-core:$log4jVersion"
-    compile "org.apache.logging.log4j:log4j-jcl:$log4jVersion"
+    compile "org.apache.logging.log4j:log4j-slf4j-impl:${log4jVersion}"
```
What is the reason to use curly braces for simple variables substitution?

NicoK replied:
imho, clearer and less error-prone in case of additions, e.g. `$log4jVersions` vs. `${log4jVersion}s`

afedulov replied:
I do not see how it is clearer. Imho this just adds the tedious part of adding the braces when pasting in a variable.

NicoK replied:
Yes, omitting the braces makes it easier to work with (selecting, then pasting etc), but that is not what I meant. I rather like to avoid ambiguities that may arise when omitting the braces, especially in a language I'm not home to (see [1] for details and a few examples like `$log4jVersion.x` looking for a property `x` there). I personally try to be consistent in the variable style (or here interpolation style) and not just use the braces when absolutely necessary.
The current style was a mix of these styles which we _could_ keep as well, but I proposed to unify that.

[1] http://docs.groovy-lang.org/latest/html/documentation/index.html#_string_interpolation

afedulov replied:
Cases like "$log4jVersion.x" are easy to avoid - the whole version has to be a variable. And unification could also go the other way, but OK.

afedulov commented on flink-job/src/main/java/com/ververica/field/dynamicrules/functions/RulesStaticJsonGenerator.java:
```diff
@@ -0,0 +1,70 @@
+      };
+
+  @Override
+  public String randomEvent(SplittableRandom rnd, long id) {
```
How is this generator intended to be used? Is is going to send the same rules with a certain frequency, like we do for the main events?

NicoK replied:
no, it's a one-time thing; no further updates

afedulov replied:
But the BaseGenerator is an 'infinitely-running' source function. Won't it be sending those updates forever?

NicoK replied:
not anymore - if `randomEvent` returns `null`, nothing will be sent. I configured `RulesStaticJsonGenerator` to wake up only once per second for not occupying too many CPU resources in that case, but kept it running to have checkpointing work (however, probably with a higher latency for sending checkpoint barriers out in that case)

afedulov replied:
Ok, I see. With the parallelism of 1 it will send all rules one by one and then stop.

NicoK replied:
well, it will stop sending updates while the source keeps running (for checkpointing purposes)

afedulov commented on flink-job/src/main/java/com/ververica/field/dynamicrules/sources/TransactionsSource.java:
```diff
@@ -61,9 +61,9 @@
       DataStream<String> transactionStrings) {
     return transactionStrings
-        .flatMap(new JsonDeserializer<Transaction>(Transaction.class))
+        .flatMap(new JsonDeserializer<>(Transaction.class))
```
Nico, did you also verify that it works as expected? I vaguely remember having to leave the full signature in certain places, because otherwise the type system strangely could not determine the types correctly (not sure if `returns()` helped).

NicoK replied:
The job is running with that change. the diamond operator should imho work everywhere since it just makes the Java compiler deduct the type

afedulov commented on flink-job/src/main/java/com/ververica/field/dynamicrules/functions/DynamicAlertFunction.java:
```diff
@@ -91,8 +91,9 @@ public void processElement(
 
-    if (noRuleAvailable(rule)) {
-      log.error("Rule with ID {} does not exist", value.getId());
+    if (rule == null) {
```
It seems that we should instead add a mitigation strategy for this occurrence instead. As I expected, this question get's raised.
http://disq.us/p/288ot93  (question from Theo D)

NicoK replied:
I would object - we're not building a product here and can leave a few things open. What you should do here, though, may even depend on the concrete use case.
I would agree to leave a `TODO` item here, if you wish.

afedulov replied:
Let's go for a comment.

afedulov commented on flink-job/src/main/java/com/ververica/field/dynamicrules/functions/RuleDeserializer.java:
```diff
@@ -29,7 +29,7 @@
 public class RuleDeserializer extends RichFlatMapFunction<String, Rule> {
 
-  private RuleParser ruleParser;
+  private transient RuleParser ruleParser;
```
Seems like IntelliJ is using some advanced AI ;)

NicoK replied:
that AI in particular sat in front of the screen ;)

afedulov commented on flink-job/src/main/java/com/ververica/field/dynamicrules/sinks/AlertsSink.java:
```diff
@@ -50,7 +51,7 @@
         return new FlinkKafkaProducer011<>(alertsTopic, new SimpleStringSchema(), kafkaProps);
       case PUBSUB:
-        return PubSubSink.<String>newBuilder()
+        return PubSubSink.newBuilder()
```
Revert. 
I am pretty sure that this does not actually work without the type when you try to use google pubsub connector.

NicoK replied:
I doubted IntelliJ for a second, but the class' generic type argument only applies to non-static members, but `org.apache.flink.streaming.connectors.gcp.pubsub.PubSubSink#newBuilder` does not have any type parameters on its own.

I don't have a google pubsub to test this against, but I suppose, **if** this needs to be added, then not for the `newBuilder` method...
(on the other hand, you kind of make the type clear by using `SimpleStringSchema`)

afedulov replied:
I do not have time  to get into the details of what is happening under the hood at the moment, but remember this failing. Unless you can test with PubSub and make sure that it works, I would propose to skip this change.

NicoK replied:
ok, let's keep it then

afedulov commented on flink-job/src/main/java/com/ververica/field/dynamicrules/sinks/CurrentRulesSink.java:
```diff
@@ -50,7 +51,7 @@
         return new FlinkKafkaProducer011<>(alertsTopic, new SimpleStringSchema(), kafkaProps);
       case PUBSUB:
-        return PubSubSink.<String>newBuilder()
+        return PubSubSink.newBuilder()
```
Revert.

afedulov commented on flink-job/src/main/java/com/ververica/field/dynamicrules/sinks/LatencySink.java:
```diff
@@ -47,7 +48,7 @@
         return new FlinkKafkaProducer011<>(latencyTopic, new SimpleStringSchema(), kafkaProps);
       case PUBSUB:
-        return PubSubSink.<String>newBuilder()
+        return PubSubSink.newBuilder()
```
Revert.

afedulov commented on flink-job/src/main/java/com/ververica/field/dynamicrules/sources/RulesSource.java:
```diff
@@ -59,7 +60,7 @@
         return kafkaConsumer;
       case PUBSUB:
-        return PubSubSource.<String>newBuilder()
+        return PubSubSource.newBuilder()
```
Revert.

afedulov commented on flink-job/src/main/java/com/ververica/field/dynamicrules/functions/TimeStamper.java:
```diff
@@ -25,7 +25,7 @@
 
   @Override
-  public void flatMap(T value, Collector<T> out) throws Exception {
+  public void flatMap(T value, Collector<T> out) {
```
I am not sure about this one. I usually like to preserve the full signature of the method that is being overridden. There is also some special handling of the exception by the framework that is underlined in the parent. Leaving it in the signature communicates that there is something to take into account here:
```
flatMap in RichFlatMapFunction:

@throws Exception This method may throw exceptions. Throwing an exception will cause the operation to fail and may trigger recovery.
```

NicoK replied:
There is nothing to take into account here though, i.e. for this function. Since it doesn't throw, it should also not advertise that (unless it is meant to be overridden again and this is an API construct; this doesn't really apply here, does it?)

afedulov replied:
I have a couple of comments:
1. This is what we do in all of our training exercises -> consistency of externally-communicated material should be something we should strive for:
https://github.com/ververica/flink-training-exercises/blob/179f940987dcf643f629051af22ce4ad67a8aa66/src/main/java/com/ververica/flinktraining/solutions/datastream_java/process/LongRidesSolution.java
2. I see no harm in being non-restrictive regarding inheritance in such public "exercises". I can see people wanting to extend our classes. By removing the throws we limit them.
3. This "throws" is in there for a reason in the framework, and this reason is documented in the comment. Leaving it reveals some important bit of information to the new users.
4. This is the default behavior of IntelliJ if you use the "Override ", and i see no good reason to fight it.

afedulov replied:
If those arguments are not convincing, feel free to remove it, I am fine with that.

NicoK replied:
Although this change isn't too important (as it doesn't change much, code-wise), I don't find your arguments convincing and would keep it.
I find it important (and actually necessary) for an API or library, but not for a concrete implementation (with the only limitation that it reduces extensibility which we do not require here).

1. actually, the new training exercises are on https://github.com/apache/flink-training and I got this change through this code as well, see https://github.com/apache/flink-training/blob/6f9ddff6e9da835ef2f526113d8a5fa98998466c/long-ride-alerts/src/solution/java/org/apache/flink/training/solutions/longrides/LongRidesSolution.java
2. we don't even publish artifacts...if people want to extend this code, they'd have to fork it anyway, making all changed they want
3. I disagree: we're not making use of that possibility from the API. It **allows throwing** an exception but this class **does not throw** any (checked) exception!
4. I'm only assuming here that you are talking about the skeleton implementation that you get when asking IntelliJ to create an override...this doesn't even have any code yet and therefore keeps everything. It does however come with a "Redundant 'throws' clause" inspection telling you this:
> Exception 'java.lang.Exception' is never thrown in the method 
 Inspection info: Reports exceptions that are declared in a method's signature but never thrown by the method itself or its implementations/derivatives.

afedulov replied:
What I see in the link you've shared is this:
```public void processElement(TaxiRide ride, Context context, Collector<TaxiRide> out) throws Exception {```

NicoK replied:
oh, yes, sorry - that one also actually may throw an exception during state updates (bad example)...this one is better: https://github.com/apache/flink-training/blob/6f9ddff6e9da835ef2f526113d8a5fa98998466c/ride-cleansing/src/solution/java/org/apache/flink/training/solutions/ridecleansing/RideCleansingSolution.java#L77

afedulov commented on flink-job/src/main/java/com/ververica/field/sources/BaseGenerator.java:
```diff
@@ -56,7 +56,7 @@ protected BaseGenerator(int maxRecordsPerSecond) {
 
   @Override
-  public void open(Configuration parameters) throws Exception {
+  public void open(Configuration parameters) {
```
Same as for the `flatMap` method exception signature in this commit.

afedulov commented on flink-job/src/main/resources/log4j2.properties:
```diff
@@ -1,7 +1,7 @@
-rootLogger.level=INFO
-rootLogger.appenderRef.console.ref=STDOUT
\ No newline at end of file
+appender.console.layout.pattern=%d{HH:mm:ss.SSS} [%t] %-5level %logger{1.} - %msg%n
```
What will be the result of using "1." as the length parameter?

NicoK replied:
it shortens the fully-qualified class name by abbreviating the package names. e.g. `o.a.f` instead of `org.apache.flink`

afedulov replied:
Ah, nice! I am for it.

afedulov commented:
Thank you for the improvements, Nico. I have left a couple of comments. I believe you have also previously mentioned some performance improvements, like better use of reflective methods access, which I could not find in those commits. Do you plan a separate PR for that?

afedulov reviewed (changes requested):
Thank you for the improvements, Nico. I have left a couple of comments. I believe you have also previously mentioned some performance improvements, like better use of reflective methods access, which I could not find in those commits. Do you plan a separate PR for that?

NicoK committed:
rework how parameter "local" works

- `local -1`: defaults, no web-UI, fixedDelayRestart
- `local <portsOrPortRange>`: defaults, web-UI on some port in the given range, fixedDelayRestart

If the web UI is started on a port range or a list of ports, the user has to
find the port in the logs from this line:
12:11:04.062 [main] INFO o.a.f.r.d.DispatcherRestEndpoint - Web frontend listening at http://localhost:8081.

afedulov commented on flink-job/src/main/java/com/ververica/field/dynamicrules/RulesEvaluator.java:
```diff
@@ -152,22 +155,31 @@ public void run() throws Exception {
-        isLocal
-            ? StreamExecutionEnvironment.createLocalEnvironmentWithWebUI(flinkConfig)
-            : StreamExecutionEnvironment.getExecutionEnvironment();
+    if (!localMode.isEmpty()) {
```
I may be missing something, but I do not see how this setup would allow to run inside of an IDE with fixed restart delay, reading from a socket and being able to work with Flink UI (this is exactly how I used 'local' and conditional restart on SOCKET source).

afedulov replied:
Ideally also without having to fix the port.

NicoK replied:
```
--local <port> --rules-source SOCKET
```
If you can live with any port, you can also provide 0 there and it will be printed in the logs (which are currently spammed by printouts though so you cannot find it before that log line is gone; that's something I have as a commit in my performance improvements)

The line you marked here though just contains the restart strategy setting for if `local` is set. The rest is a few lines above and sockets as sources at a completely different place as before.

afedulov replied:
What I mean is if you want to have Flink UI in this setup, you cannot set a fixed delay restarts. This is what is needed, because otherwise the flood of Flink messages overwhelms the IDE if the socket happened to be closed.
